x-pack/plugin/async-search/src/main/java/org/elasticsearch/xpack/search/AsyncSearchTask.java

```diff
@@ -220,38 +218,45 @@ private void internalAddCompletionListener(ActionListener<AsyncSearchResponse> l
                 // ensure that we consumes the listener only once
                 AtomicBoolean hasRun = new AtomicBoolean(false);
                 long id = completionId++;
-
                 final Cancellable cancellable;
                 try {
-                    cancellable = threadPool.schedule(() -> {
-                        if (hasRun.compareAndSet(false, true)) {
-                            // timeout occurred before completion
-                            removeCompletionListener(id);
-                            listener.onResponse(getResponseWithHeaders());
-                        }
-                    }, waitForCompletion, "generic");
-                } catch (EsRejectedExecutionException exc) {
+                     cancellable = threadPool.schedule(
+                         () -> {
+                            if (hasRun.compareAndSet(false, true)) {
+                                // timeout occurred before completion
+                                removeCompletionListener(id);
+                                getResponseWithHeaders(listener);
+                            }
+                        },
+                        waitForCompletion,
+                        "generic");
+                } catch(Exception exc) {
                     listener.onFailure(exc);
                     return;
                 }
-                completionListeners.put(id, resp -> {
-                    if (hasRun.compareAndSet(false, true)) {
-                        // completion occurred before timeout
-                        cancellable.cancel();
-                        listener.onResponse(resp);
-                    }
-                });
+                completionListeners.register(
+                    id,
+                    ActionListener.wrap(
+                        resp -> {
+                            if (hasRun.compareAndSet(false, true)) {
+                                // completion occurred before timeout
+                                cancellable.cancel();
+                                listener.onResponse(resp);
+                            }
+                        },
+                        listener::onFailure
```

Comment:
we still need to cancel the cancellable on failure

Reply:
yes indeed :)